jupyterhub/handlers/base.py

```diff
@@ -808,7 +808,12 @@ async def auth_to_user(self, authenticated, user=None):
             group_names = authenticated.get("groups")
             if group_names is not None:
                 user.sync_groups(group_names)
-
+        # apply authenticator-managed roles
+        if self.authenticator.manage_roles:
+            print(authenticated)
+            auth_roles = authenticated.get("roles")
+            if auth_roles is not None:
+                user.sync_roles(auth_roles)
```

Comment:
I infer that the original intent of this PR was to have something like:

```python
{
   'auth_state': ...,
   'name': 'elephant',
   'groups': ['animals'],
   'roles': [
      {
          'name': 'custom-role',
          'description': 'My custom role',
          'scopes': [...],
          'groups': ['animals']
      }
   ]
}
```

Bu that assumes that both `groups` and `roles` will be in sync and always returned together. Even for that use case the current logic still has the downside that if `custom-role` is extended to cover `['animals', 'fungi']` it would not be reflected without restart.

I also think the API should be designed with the least surprise principle, so that the `roles` are explicitly added for the current user.

I do not know if `'users`' key should be disallowed in the managed roles. It feels to me that it is not needed in this context. Thoughts?